clustered indexes: add initial docs (#4476)
pingcap/docs: clustered-indexes.md

---
title: Clustered Indexes
summary: Learn how clustered indexes apply to TiDB.
---

# Clustered Indexes

> The clustered index is an experimental feature introduced in TiDB 5.0-rc. This document provides multiple examples to explain how this feature makes a difference to the query performance of TiDB. To enable this feature and see the detailed operation guide, see [tidb_enable_clustered_index](/system-variables.md#tidb_enable_clustered_index).

Reviewer: ```suggestion
The clustered index is an experimental feature introduced in TiDB 5.0.0-rc. This document provides multiple examples to explain how this feature makes a difference to the query performance of TiDB. To enable this feature and see the detailed operation guide, see [tidb_enable_clustered_index](/system-variables.md#tidb_enable_clustered_index).
```

Clustered indexes provide TiDB the ability to organize tables in a way that can improve the performance of certain queries. The term _clustered_ in this context refers to the _organization of how data is stored_ and not _a group of database servers working together_. Some database management systems refer to clustered indexes as _index-organized tables_ (IOT).

> TiDB supports clustering only by a table's `PRIMARY KEY`. With clustered indexes enabled, the terms _the_ `PRIMARY KEY` and _the clustered index_ might be used interchangeably. `PRIMARY KEY` refers to the constraint (a logical property), and clustered index describes the physical implementation of how the data is stored.

Reviewer: "the terms _ `PRIMARY KEY`_ and _clustered index_ might be used interchangeably. " This is a bit confusing. How can two terms used interchangeably? Could you change the wording?
```suggestion
TiDB supports data clustering only by a table's `PRIMARY KEY`. With clustered indexes enabled, the terms _ `PRIMARY KEY`_ and _clustered index_ might be used interchangeably. `PRIMARY KEY` refers to the constraint (a logical property), and clustered index describes the physical implementation of how the data is stored.
```

Author: The sentence that follows explains how they can be used interchangebly. Some databases permit you to chose which index to cluster by, but TiDB does not. Some also permit clustering by moe than one index (such as TokuDB), but TiDB does not. Because TiDB has one clustered index (_the clustered index_) it is always the same index as the `PRIMARY KEY`.

I would prefer without the edits. Data clustering is redundant (clustering already refers to data) and removing "the" makes it less correct for the case with databases that support multiple clustered indexes.

> ## Default behavior

Reviewer: ```suggestion
## Limited support before v5.0
```

TiDB has always supported clustered indexes, provided the following criteria are true:

- The table contains a `PRIMARY KEY`
- The primary key is an `INTEGER` or `BIGINT`

> - The primary key consists of only one column

Reviewer: Any difference between `PRIMARY KEY` and primary key?

Author: No, it's the same.

```

```sql
Query OK, 0 rows affected (0.09 sec)

Query OK, 0 rows affected (0.10 sec)

Records: 2  Duplicates: 0  Warnings: 0

Records: 2  Duplicates: 0  Warnings: 0

+-------------+---------+------+---------------------------------------+---------------+
| id          | estRows | task | access object                         | operator info |
+-------------+---------+------+---------------------------------------+---------------+
| Point_Get_1 | 1.00    | root | table:always_clusters_in_all_versions | handle:1      |
+-------------+---------+------+---------------------------------------+---------------+
1 row in set (0.00 sec)

+-------------+---------+------+--------------------------------------------------------+---------------+
| id          | estRows | task | access object                                          | operator info |
+-------------+---------+------+--------------------------------------------------------+---------------+
| Point_Get_1 | 1.00    | root | table:does_not_cluster_by_default, index:PRIMARY(guid) |               |
+-------------+---------+------+--------------------------------------------------------+---------------+
1 row in set (0.00 sec)
```

> The `EXPLAIN` results look similar, but in the second example TiDB must first read the `PRIMARY KEY` index on `guid` in order to find the `handle` value. This is more obvious in the following example where the value of the `PRIMARY KEY` is not in the index on `does_not_cluster_by_default.b`. An extra lookup must be performed on the table rows (`└─TableFullScan_5`) to convert the `handle` to the `PRIMARY KEY` value of `guid`:

Reviewer: ```suggestion
The two `EXPLAIN` results above look similar, but in the second example, TiDB must first read the `PRIMARY KEY` index on the `guid` column in order to find the `handle` value. This is more obvious in the following example where the `PRIMARY KEY` value is not in the index on `does_not_cluster_by_default.b`. TiDB must perform an extra full table scan on the table rows (`└─TableFullScan_5`) to convert the `handle` value to the `PRIMARY KEY` value of `guid`:
```

Author: I don't have permissions to partially accept your edit, but it's not correct. It is an additional lookup, not an extra full table scan.

Reviewer: Comment changed. PTAL again, thanks~

3 rows in set (0.01 sec)

+---------------------------+---------+-----------+-----------------------------------+------------------------------------------------+
| id                        | estRows | task      | access object                     | operator info                                  |
+---------------------------+---------+-----------+-----------------------------------+------------------------------------------------+
| Projection_4              | 0.00    | root      |                                   | test.does_not_cluster_by_default.guid          |
| └─TableReader_7           | 0.00    | root      |                                   | data:Selection_6                               |
|   └─Selection_6           | 0.00    | cop[tikv] |                                   | eq(test.does_not_cluster_by_default.b, "aaaa") |
|     └─TableFullScan_5     | 2.00    | cop[tikv] | table:does_not_cluster_by_default | keep order:false, stats:pseudo                 |
+---------------------------+---------+-----------+-----------------------------------+------------------------------------------------+
4 rows in set (0.00 sec)
```

> # With clustered indexes

Reviewer: The previous section also discusses clustered indexes ("TiDB has always supported clustered indexes, provided the following criteria are true:"). So this title is a bit confusing.

Author: Agreed. Can we change it to "Changes since in TiDB 5.0"?

> # Changes since TiDB 5.0

Reviewer: ```suggestion
# Full support since TiDB v5.0
```

> New to TiDB 5.0-rc, clustering is permitted by any `PRIMARY KEY`. The following `EXPLAIN` output shows the previous example with clustered indexes now enabled:

Reviewer: ```suggestion
Since v5.0, TiDB supports clustering by any `PRIMARY KEY`. The following `EXPLAIN` output shows the previous example with clustered indexes enabled:
```

> Since v5.0, TiDB supports clustering by any `PRIMARY KEY`. The following `EXPLAIN` output shows the previous example with clustered indexes enabled:

Reviewer: Is this the difference between the clustered indexes before v5.0 (as the previous section describes) and the clustered indexes for v5.0?

Author: yes.

Reviewer: ```suggestion
Since v5.0, TiDB provides full support for clustered indexes by any `PRIMARY KEY`. The following `EXPLAIN` output shows the previous example with clustered indexes enabled:
```

+-------------+---------+------+-------------------------------------------------------+---------------+
1 row in set (0.00 sec)

+--------------------------+---------+-----------+------------------------------------+-------------------------------------------------------+
| id                       | estRows | task      | access object                      | operator info                                         |
+--------------------------+---------+-----------+------------------------------------+-------------------------------------------------------+
| Projection_4             | 10.00   | root      |                                    | test.will_now_cluster.guid                            |
| └─IndexReader_6          | 10.00   | root      |                                    | index:IndexRangeScan_5                                |
|   └─IndexRangeScan_5     | 10.00   | cop[tikv] | table:will_now_cluster, index:b(b) | range:["aaaa","aaaa"], keep order:false, stats:pseudo |
+--------------------------+---------+-----------+------------------------------------+-------------------------------------------------------+
3 rows in set (0.00 sec)
```

> Clustering by a composite `PRIMARY KEY` is also supported:

Reviewer: Can we use `clustered index` instead?

Author: Yes, you could say "A clustered index on a composite `PRIMARY KEY`".

Reviewer 2: `Cluster by ...` is used a lot in this document. It's clear and I suggest we keep it. @TomShawn

> Clustering by a composite `PRIMARY KEY` is also supported:

Reviewer: ```suggestion
Clustered index by a composite `PRIMARY KEY` is also supported:
```

Author: I have used clustering earlier in the document already. It refers to "organizing data" by a composite key.

```sql
SET tidb_enable_clustered_index = 1;
CREATE TABLE composite_primary_key (
 key_a INT NOT NULL,
 key_b INT NOT NULL,
 b CHAR(100),
 PRIMARY KEY (key_a, key_b)
);

INSERT INTO composite_primary_key VALUES (1, 1, 'aaa'), (2, 2, 'bbb');
EXPLAIN SELECT * FROM composite_primary_key WHERE key_a = 1 AND key_b = 2;
```

```sql
Query OK, 0 rows affected (0.00 sec)

Query OK, 0 rows affected (0.09 sec)

Query OK, 2 rows affected (0.02 sec)
Records: 2  Duplicates: 0  Warnings: 0

+-------------+---------+------+--------------------------------------------------------------------+---------------+
| id          | estRows | task | access object                                                      | operator info |
+-------------+---------+------+--------------------------------------------------------------------+---------------+
| Point_Get_1 | 1.00    | root | table:composite_primary_key, clustered index:PRIMARY(key_a, key_b) |               |
+-------------+---------+------+--------------------------------------------------------------------+---------------+
1 row in set (0.00 sec)
```

> This behavior is consistent with MySQL, where the InnoDB storage engine will by default cluster by any `PRIMARY KEY`.

Reviewer: Global comment: Could you explain a bit what "clustering" means? Does it mean the clustered index? If not, what's the subject or object of clustering?

Author: There is a sentence in the intro which covers this:

> The term _clustered_ in this context refers to the _organization of how data is stored_ and not _a group of database servers working together_

Reviewer: ```suggestion
This behavior is consistent with MySQL, where the data on the InnoDB storage engine will by default cluster by any `PRIMARY KEY`.
```

Author: I don't think this is neccessary. Storage engines only store data, so it is implied.

> ## Storage considerations

Reviewer: ```suggestion
## Influences on storage
```

Author: I prefer storage considerations, since the feature is usually helpful but you need to consider if the storage cost is worth it.

> Because the `PRIMARY KEY` replaces a 64-bit `handle` as the internal pointer to table rows, using clustered indexes might increase storage requirements. This is particularly impactful on tables that contain many secondary indexes. Consider the following example:

Reviewer: ```suggestion
Because the `PRIMARY KEY` replaces a 64-bit `handle` value as the internal pointer to table rows, using clustered indexes might increase storage requirements. This is particularly impactful on tables that contain many secondary indexes. Consider the following example:
```

```sql
CREATE TABLE t1 (
 guid CHAR(32) NOT NULL PRIMARY KEY,
 b BIGINT,
 INDEX(b)
);
```

> Since the pointer to the `guid` is a `char(32)`, each index value for `b` will now require approximately `8 + 32 = 40 bytes` (a `BIGINT` value requires 8 bytes for storage). This compares to `8 + 8 = 16 bytes` for non-clustered tables. The exact storage requirements will differ after compression has been applied.

Reviewer: ```suggestion
Because the pointer to the `guid` is a `char(32)`, each index value for `b` will now require approximately `8 + 32 = 40 bytes` (a `BIGINT` value requires 8 bytes for storage). This compares to `8 + 8 = 16 bytes` for non-clustered tables. The exact storage requirements will differ after compression has been applied.
```

---
title: Clustered Indexes
summary: Learn how clustered indexes apply to TiDB.
---

# Clustered Indexes

> **Warning:**
>

> > Clustered indexes is an experimental feature introduced in TiDB 5.0.

Reviewer 3: ```suggestion
> Clustered indexes is an experimental feature introduced in TiDB 5.0-rc.
```
So far as I know it will be GA in TiDB 5.0, it is experimental only in 5.0-rc.

> > Clustered indexes is an experimental feature introduced in TiDB 5.0-rc.

Reviewer: The clustered index is an experimental feature introduced in TiDB 5.0-rc. This document provides multiple examples to explain how this feature makes a difference to the query performance of TiDB. To enable this feature and see the detailed operation guide, see [tidb_enable_clustered_index
](/system-variables.md#tidb_enable_clustered_index-).

> Clustered indexes provide the ability to organize tables in a way that can improve the performance of certain queries. The term _clustered_ in this context refers to the _organization of how data is stored_ and not _a group of database servers working together_. Some database management systems refer to clustered indexes as _index-organized tables_ (IOT).

Reviewer: ```suggestion
Clustered indexes provide TiDB the ability to organize tables in a way that can improve the performance of certain queries. The term _clustered_ in this context refers to the _organization of how data is stored_ and not _a group of database servers working together_. Some database management systems refer to clustered indexes as _index-organized tables_ (IOT).
```

> TiDB only supports clustering by a table's `PRIMARY KEY`. With clustered indexes enabled, the terms _the_ `PRIMARY KEY` and _the clustered index_ might be used interchangeably.

Author: I wanted to say this because I think this is a concept that confuses new users. But I think the wording may still be a little clumsy (we use "might" in docs instead of "may" to avoid ambiguity, but it doesn't sound that natural to me.)

The `PRIMARY KEY` refers to the constraint versus clustered index refering to the physical implementation. It's not the same thing, but yes, in MySQL they are often used interchangeably.

## Default behavior

> TiDB has always supported clustered indexes, provided the following criteria are true:

Author: I assume since before 1.0 (it has since I started using it). The differences are hilighted - in the situation where the _folllowing criteria are true_ TiDB 5.0 will not change anything for you.

> TiDB has always supported clustered indexes, provided the following criteria are true:

Reviewer: @coocood @bb7133 Could you help confirm when is `always`?

> TiDB has always supported clustered indexes, provided the following criteria are true:

Author: That implies that it is a different type of clustered index, which is not the case. It has always clustered indexes provided certain criteria are true. Now it will cluster in all scenarios.

> TiDB has always supported clustered indexes, provided the following criteria are true:

Reviewer: ```suggestion
Before v5.0, TiDB has only limited support for clustered indexes, provided the following criteria are true:
```

3 rows in set (0.01 sec)

+---------------------------+---------+-----------+-----------------------------------+------------------------------------------------+
| id                        | estRows | task      | access object                     | operator info                                  |
+---------------------------+---------+-----------+-----------------------------------+------------------------------------------------+
| Projection_4              | 0.00    | root      |                                   | test.does_not_cluster_by_default.guid          |
| └─TableReader_7           | 0.00    | root      |                                   | data:Selection_6                               |
|   └─Selection_6           | 0.00    | cop[tikv] |                                   | eq(test.does_not_cluster_by_default.b, "aaaa") |
|     └─TableFullScan_5     | 2.00    | cop[tikv] | table:does_not_cluster_by_default | keep order:false, stats:pseudo                 |
+---------------------------+---------+-----------+-----------------------------------+------------------------------------------------+
4 rows in set (0.00 sec)
```

# With clustered indexes

> New to TiDB 5.0, clustering is permitted by any `PRIMARY KEY`. The following `EXPLAIN` output shows the previous example with clustered indexes now enabled:

Author: ```suggestion
New to TiDB 5.0-rc, clustering is permitted by any `PRIMARY KEY`. The following `EXPLAIN` output shows the previous example with clustered indexes now enabled:
```
